### PR TITLE
[EASI-3712] Update model type options (backend)

### PR DESCRIPTION
**Pull request by OddTomBrooks**
# EASI-3712

## Changes and Description

- GQL
- SQL
  - Migration
  - Store
- Go
  - Data model
  - Unit test updates

## How to test this change

- disable this PR's flyway migration
- bring up fresh containers
- insert data using the sql provided below
- enable this PR's flyway migration
- assert that the database contains appropriate values
- run unit tests

```
INSERT INTO "public"."user_account"("id", "username", "is_euaid", "common_name", "locale", "email", "given_name", "family_name", "zone_info", "has_logged_in") VALUES('762d3239-8dca-4486-8a7b-fe6255eab1e8', 'MINT', 'TRUE', 'MINT Doe', 'en_US', 'MINT@local.cms.gov', 'MINT', 'Doe', 'America/Los_Angeles', 'TRUE');


/* Voluntary Type */
INSERT INTO "public"."model_plan"("id", "model_name", "archived", "status", "created_dts", "created_by") VALUES('0940d64f-17bc-4f33-ad7c-c82a08f1b19c', 'Excellent Model', 'FALSE', 'PLAN_DRAFT', '2023-12-19 15:32:57.692596+00', '762d3239-8dca-4486-8a7b-fe6255eab1e8');
INSERT INTO "public"."plan_basics"("id", "model_plan_id", "model_type", "created_dts", "status", "created_by") VALUES('a904f2d6-3d26-4202-b67e-328186257000', '0940d64f-17bc-4f33-ad7c-c82a08f1b19c', 'VOLUNTARY', '2023-12-19 15:32:57.692596+00', 'READY', '762d3239-8dca-4486-8a7b-fe6255eab1e8');

/* Mandatory Type */
INSERT INTO "public"."model_plan"("id", "model_name", "archived", "status", "created_dts", "created_by") VALUES('0940d64f-17bc-4f33-ad7c-c82a08f1b19d', 'Excellent Model', 'FALSE', 'PLAN_DRAFT', '2023-12-19 15:32:57.692596+00', '762d3239-8dca-4486-8a7b-fe6255eab1e8');
INSERT INTO "public"."plan_basics"("id", "model_plan_id", "model_type", "created_dts", "status", "created_by") VALUES('a904f2d6-3d26-4202-b67e-328186257001', '0940d64f-17bc-4f33-ad7c-c82a08f1b19d', 'MANDATORY', '2023-12-19 15:32:57.692596+00', 'READY', '762d3239-8dca-4486-8a7b-fe6255eab1e8');


/* TBD Type */
INSERT INTO "public"."model_plan"("id", "model_name", "archived", "status", "created_dts", "created_by") VALUES('0940d64f-17bc-4f33-ad7c-c82a08f1b19e', 'Excellent Model', 'FALSE', 'PLAN_DRAFT', '2023-12-19 15:32:57.692596+00', '762d3239-8dca-4486-8a7b-fe6255eab1e8');
INSERT INTO "public"."plan_basics"("id", "model_plan_id", "model_type", "created_dts", "status", "created_by") VALUES('a904f2d6-3d26-4202-b67e-328186257002', '0940d64f-17bc-4f33-ad7c-c82a08f1b19e', 'TBD', '2023-12-19 15:32:57.692596+00', 'READY', '762d3239-8dca-4486-8a7b-fe6255eab1e8')
```

## PR Author Review Checklist

- [x] Met the ticket's acceptance criteria, or will meet them in a subsequent PR.
- [x] Added or updated tests for backend resolvers or other functions as needed.
- [ ] Added or updated client tests for new components, parent components, functions, or e2e tests as necessary.
- [x] Updated the [Postman Collection](../MINT.postman_collection.json) if necessary.

## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.

### CODE REVIEW

File: pkg/graph/resolvers/plan_basics_test.go
Comment (OddTomBrooks) [inline pkg/graph/resolvers/plan_basics_test.go:92]: These two lines are the newly added ones. The other lines are indented to preserve formatting.

**Comment by OddTomBrooks** [comment]
Linter and client issues are related to unresolved frontend requirements, pending be+fe merge

File: migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql
Comment (StevenWadeOddball) [inline migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql:16]: @OddTomBrooks, was `POSTGres` complaining if you didn't add a commit here? There might be a way around using it. 

I'm a bit uneasy with the commit in the middle. Flyway calls each of these scripts in their own transaction, and rolls them back if there are any error. I don't know what happens if you have already committed partially into the script (does those changes stay?).

I think it would be better to remove the commit.

File: migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql
Comment (StevenWadeOddball) [inline migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql:30]: @OddTomBrooks, are these options going to be manually updated in PROD? This is what I see in JIRA. 

I don't see instructions for how to handle `TBD`, but your approach seems logical. Did you confirm this with product?

```
Historical data migration for those that had "Mandatory" selected:
CJR 3-Year Extension << update to Mandatory regional or state
Expanded Home Health Value-Based Purchasing Model << update to Mandatory national
Increasing Organ Transplant Access (IOTA) Model << update to Mandatory regional or state
Most Favored Nation Model (MFN) << update to Mandatory national

```

File: migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql
Comment (StevenWadeOddball) [inline migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql:37]: This doesn't work, this removes all data in the model type column

File: migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql
Comment (StevenWadeOddball) [inline migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql:26]: These updates will fail if you don't set , modified columns

File: migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql
Comment (StevenWadeOddball) [inline migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql:31]: same about needing to set modified columns

**Review by StevenWadeOddball — CHANGES_REQUESTED** [review]
Thanks @OddTomBrooks, app code looks good, the feedback is largely around the migration. From my testing, it isn't working as anticipated. It drops all the data in those columns. 

I think for best practice, these should be manually tested and validated to ensure the migration is behaving as expected.  ( I suggest something like I did here)
1. Create test data with SQL
2. Run all migrations up to the new migration
3. Confirm your test cases

This revealed to me that modified columns need to be handled, and that the data gets removed from the model_type column (instead of being added to an array).

Here is my test SQL if you want to use that to test
```SQL
INSERT INTO "public"."user_account"("id", "username", "is_euaid", "common_name", "locale", "email", "given_name", "family_name", "zone_info", "has_logged_in") VALUES('762d3239-8dca-4486-8a7b-fe6255eab1e8', 'MINT', 'TRUE', 'MINT Doe', 'en_US', 'MINT@local.cms.gov', 'MINT', 'Doe', 'America/Los_Angeles', 'TRUE');


/* Voluntary Type */
INSERT INTO "public"."model_plan"("id", "model_name", "archived", "status", "created_dts", "created_by") VALUES('0940d64f-17bc-4f33-ad7c-c82a08f1b19c', 'Excellent Model', 'FALSE', 'PLAN_DRAFT', '2023-12-19 15:32:57.692596+00', '762d3239-8dca-4486-8a7b-fe6255eab1e8');
INSERT INTO "public"."plan_basics"("id", "model_plan_id", "model_type", "created_dts", "status", "created_by") VALUES('a904f2d6-3d26-4202-b67e-328186257000', '0940d64f-17bc-4f33-ad7c-c82a08f1b19c', 'VOLUNTARY', '2023-12-19 15:32:57.692596+00', 'READY', '762d3239-8dca-4486-8a7b-fe6255eab1e8'); 

/* Mandatory Type */
INSERT INTO "public"."model_plan"("id", "model_name", "archived", "status", "created_dts", "created_by") VALUES('0940d64f-17bc-4f33-ad7c-c82a08f1b19d', 'Excellent Model', 'FALSE', 'PLAN_DRAFT', '2023-12-19 15:32:57.692596+00', '762d3239-8dca-4486-8a7b-fe6255eab1e8');
INSERT INTO "public"."plan_basics"("id", "model_plan_id", "model_type", "created_dts", "status", "created_by") VALUES('a904f2d6-3d26-4202-b67e-328186257001', '0940d64f-17bc-4f33-ad7c-c82a08f1b19d', 'MANDATORY', '2023-12-19 15:32:57.692596+00', 'READY', '762d3239-8dca-4486-8a7b-fe6255eab1e8');


/* TBD Type */
INSERT INTO "public"."model_plan"("id", "model_name", "archived", "status", "created_dts", "created_by") VALUES('0940d64f-17bc-4f33-ad7c-c82a08f1b19e', 'Excellent Model', 'FALSE', 'PLAN_DRAFT', '2023-12-19 15:32:57.692596+00', '762d3239-8dca-4486-8a7b-fe6255eab1e8');
INSERT INTO "public"."plan_basics"("id", "model_plan_id", "model_type", "created_dts", "status", "created_by") VALUES('a904f2d6-3d26-4202-b67e-328186257002', '0940d64f-17bc-4f33-ad7c-c82a08f1b19e', 'TBD', '2023-12-19 15:32:57.692596+00', 'READY', '762d3239-8dca-4486-8a7b-fe6255eab1e8')
```

File: migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql
Comment (OddTomBrooks) [reply migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql:30]: Yes exactly. Gary and I discussed this and I'll reach out to Jeff for confirmation.

File: migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql
Comment (OddTomBrooks) [reply migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql:16]: Yes exactly. There was an error trying to use the enums before committing, otherwise it appears the choice is to separate this into two scripts. I'll take another look at this, I remember running into a similar issue before some time ago and I think the resolution was just to make two flyway migrations.

Comment (StevenWadeOddball) [reply migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql:16]: 👍 , I have also gotten around this before by casting the column to TEXT, doing the update, and then casting to the new type. That should likely resolve the issue, and make sure the whole migration is reverted instead of separating into two migrations.

Comment (OddTomBrooks) [reply migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql:16]: This makes sense. Do you remember the migration? If not, it should be fine, I just want to maintain consistency.

Comment (StevenWadeOddball) [reply migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql:16]: Sure! It's in main , V122__Update_Funding_Source_Type.sql

I set the column type to TEXT, then when I'm doing the update, the new values are just text. Then after the text is updated, I cast it to the new type

File: migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql
Comment (OddTomBrooks) [reply migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql:17]: I reached out to Jeff regarding this. He instructed that from a product perspective it's best to simply drop the TBD option, so we are nulling it out.

File: migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql
Comment (OddTomBrooks) [reply migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql:31]: Resolved

File: migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql
Comment (StevenWadeOddball) [reply migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql]: Thanks Tom, also to confirm, are the individual models that need to  be `MANDATORY_NATIONAL` being manually updated? (And who is doing that if so?)

**Comment by StevenWadeOddball** [comment]
Quick note, the files changed are off in this branch (from pulling main directly into this branch instead of the feature). Can you please update the feature branch with main, and pull it into this branch? I believe that should correct this showing 81 files changed

**Comment by OddTomBrooks** [comment]
> Quick note, the files changed are off in this branch (from pulling main directly into this branch instead of the feature). Can you please update the feature branch with main, and pull it into this branch? I believe that should correct this showing 81 files changed

Ah yes. Apologies, pulled in main to update the migration version so you didn’t have to worry about that, didn’t think about the file expansion. You are correct that merging main into the feature branch will simplify the changed files.

File: migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql
Comment (OddTomBrooks) [reply migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql]: I am doing that manually via database edits once the merge goes through. I didn’t feel the specificity of model names as a migration policy or sharing model data in GitHub through a migration was appropriate. It’s only a few rows, and I minimized manual audits by having a default case migration.

Comment (OddTomBrooks) [reply migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql]: By merge, I mean once the feature is deployed.

**Comment by StevenWadeOddball** [comment]
> > Quick note, the files changed are off in this branch (from pulling main directly into this branch instead of the feature). Can you please update the feature branch with main, and pull it into this branch? I believe that should correct this showing 81 files changed
> 
> Ah yes. Apologies, pulled in main to update the migration version so you didn’t have to worry about that, didn’t think about the file expansion. You are correct that merging main into the feature branch will simplify the changed files.

Thanks @OddTomBrooks, I updated the feature branch, but the diff is still showing the other files changed. Because we use squash and merge, I suspect this branch needs to pull in the changes from the feature branch before doing a squash and merge (or it will show that these files were changed here instead of correctly resolving the  diffs). 

My main concern is just showing an incorrect GIT history for when the files were last changed etc.

File: migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql
Comment (StevenWadeOddball) [reply migrations/V122__Update_PlanBasics_ModelType_MultiSelect_And_Other_Text.sql]: Perfect, yeah that works for me. I just wanted to make sure I understood what the approach is. But I agree, these can be handled in a one off manor. It might be less complex just to resolve those few in the UI

**Comment by OddTomBrooks** [comment]
> > > Quick note, the files changed are off in this branch (from pulling main directly into this branch instead of the feature). Can you please update the feature branch with main, and pull it into this branch? I believe that should correct this showing 81 files changed
> > 
> > 
> > Ah yes. Apologies, pulled in main to update the migration version so you didn’t have to worry about that, didn’t think about the file expansion. You are correct that merging main into the feature branch will simplify the changed files.
> 
> Thanks @OddTomBrooks, I updated the feature branch, but the diff is still showing the other files changed. Because we use squash and merge, I suspect this branch needs to pull in the changes from the feature branch before doing a squash and merge (or it will show that these files were changed here instead of correctly resolving the diffs).
> 
> My main concern is just showing an incorrect GIT history for when the files were last changed etc.

It’s not a problem. I’ll just revert back to HEAD~2 and reupdate the migration version. This is an interesting nuance to keep in mind as we move into a feature branch oriented workflow.

**Review by StevenWadeOddball — COMMENTED** [review]
Thanks @OddTomBrooks , I validated the migration changes and they look good, and are working as expected. 

Holding off on approval just for GIT history. Can you merge the feature into this branch please? It's showing a lot of changes from main as new changes in your branch.

**Review by StevenWadeOddball — CHANGES_REQUESTED** [review]
@garyjzhao , It looks like there is an issue in the front end. 

The readview isn't displaying the other value correctly
image


Read View
image



Can you please correct this? Looks like a snapshot might be failing too. CC @patrickseguraoddball

**Comment by StevenWadeOddball** [comment]
Note, @OddTomBrooks , the backend changes look good, just waiting for the FE for final approval now

**Comment by OddTomBrooks** [comment]
> Note, @OddTomBrooks , the backend changes look good, just waiting for the FE for final approval now

Nice. Thank you.

**Review by StevenWadeOddball — APPROVED** [review]
Thanks @garyjzhao , the changes look good!
image

**Pull request merged**
